### ISSUE TITLE
no-constant-condition false positive

### ISSUE THREAD

**Issue by MethodGrab** [eslint/eslint#4737]
eslint: v1.10.3
Config: `{"rules": { "no-constant-condition": 2, "no-array-constructor": 2 }}`

**Problem:**
`no-constant-condition` gives a false positive result for the following code:
```js
if ( 'every' in [] ) { /* browser supports every */ }
```
Using `Array()` instead of `[]` will not result in the error but I have `no-array-constructor` so it still doesn't pass linting.

**Expected:**
The results for `[]` and `Array()` are consistent, linting passes.

**Actual:**
Linting fails.

**Example Code:**
```js
if ( 'every' in [] ) { console.log( 1 ) } // → no-constant-condition error

if ( 'every' in Array() ) { console.log( 2 ) } // → no-array-constructor error

/* eslint-disable no-array-constructor */
if ( 'every' in Array() ) { console.log( 3 ) } // → pass
/* eslint-enable no-array-constructor */

if ( typeof [].every === 'function' ) { console.log( 4 ) } // → pass
```

**ESLint Output:**
```
1:1   error  Unexpected constant condition                 no-constant-condition
3:17  error  The array literal notation [] is preferrable  no-array-constructor
```

**Comment by ilyavolodin** [comment]
Looks like a bug. Not really surprised, `in` operator is used very infrequently and we probably missed it in the rule all together.

**Comment by michaelficarra** [comment]
But it *is* constant in a particular environment, assuming you're not mutating native prototypes. You should just disable the rule here.

**Comment by ilyavolodin** [comment]
@michaelficarra You are absolutely correct about particular environment, however, JavaScript is executed in a lot of different environments (in this case "environment" is a broader term that includes different version of browsers). A specific property on the object might be available in one browser, but not in another, so the expression is not constant in cross-environment code.

**Comment by michaelficarra** [comment]
So code like `if ("v" === "\v") ... else ...` should not be warned either because some environments take one path and some take another? Eslint sticks to standard JS, not implementations.

**Comment by nzakas** [comment]
I agree with @ilyavolodin here. The intent of this rule is to flag conditions that will never change regardless of where code is executed (see the examples in the documentation).

**Comment by ilyavolodin** [comment]
@michaelficarra You example is correct, but extreme. We have to find middle ground, where the rule is helpful to users, but at the same time doesn't cover all possible variations. In this case, it's a well known fact that different versions of browsers have different attributes for native objects. Forking code based on existence of those attributes is a common practice. While `"v" === "\v"` might not be true in some cases, those are very much edge cases.

**Comment by MethodGrab** [comment]
Thanks for fixing this so quickly!
